shared/src/commonMain/kotlin/org/hyperskill/app/challenges/widget/presentation/ChallengeWidgetActionDispatcher.kt

```diff
@@ -0,0 +1,132 @@
+package org.hyperskill.app.challenges.widget.presentation
+
+import kotlin.time.DurationUnit
+import kotlin.time.toDuration
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import org.hyperskill.app.analytic.domain.interactor.AnalyticInteractor
+import org.hyperskill.app.challenges.domain.repository.ChallengesRepository
+import org.hyperskill.app.challenges.widget.presentation.ChallengeWidgetFeature.Action
+import org.hyperskill.app.challenges.widget.presentation.ChallengeWidgetFeature.InternalAction
+import org.hyperskill.app.challenges.widget.presentation.ChallengeWidgetFeature.InternalMessage
+import org.hyperskill.app.challenges.widget.presentation.ChallengeWidgetFeature.Message
+import org.hyperskill.app.core.presentation.ActionDispatcherOptions
+import org.hyperskill.app.magic_links.domain.interactor.MagicLinksInteractor
+import org.hyperskill.app.sentry.domain.interactor.SentryInteractor
+import org.hyperskill.app.sentry.domain.model.transaction.HyperskillSentryTransactionBuilder
+import org.hyperskill.app.sentry.domain.withTransaction
+import org.hyperskill.app.step_completion.domain.flow.DailyStepCompletedFlow
+import org.hyperskill.app.step_completion.domain.flow.TopicCompletedFlow
+import ru.nobird.app.presentation.redux.dispatcher.CoroutineActionDispatcher
+
+class ChallengeWidgetActionDispatcher(
+    config: ActionDispatcherOptions,
+    solvedStepsSharedFlow: SharedFlow<Long>,
+    topicCompletedFlow: TopicCompletedFlow,
+    dailyStepCompletedFlow: DailyStepCompletedFlow,
+    private val challengesRepository: ChallengesRepository,
+    private val magicLinksInteractor: MagicLinksInteractor,
+    private val sentryInteractor: SentryInteractor,
+    private val analyticInteractor: AnalyticInteractor
+) : CoroutineActionDispatcher<Action, Message>(config.createConfig()) {
+    private var timerJob: Job? = null
+
+    companion object {
+        private val TIMER_TICK_INTERVAL = 1.toDuration(DurationUnit.MINUTES)
+    }
+
+    init {
+        solvedStepsSharedFlow
+            .distinctUntilChanged()
+            .onEach {
+                onNewMessage(InternalMessage.StepSolved)
+            }
+            .launchIn(actionScope)
+
+        dailyStepCompletedFlow.observe()
+            .distinctUntilChanged()
+            .onEach {
+                onNewMessage(InternalMessage.DailyStepCompleted)
+            }
+            .launchIn(actionScope)
+
+        topicCompletedFlow.observe()
+            .distinctUntilChanged()
+            .onEach {
+                onNewMessage(InternalMessage.TopicCompleted)
+            }
+            .launchIn(actionScope)
```

Comment:
@ivan-magda it seems that we don't cover all necessary cases here. If user completes stage in mobile apps the challenge card won't be updated, is it ok?

Reply:
@vladkash we are handling this, see for more info
https://github.com/hyperskill/mobile-app/blob/develop/shared/src/commonMain/kotlin/org/hyperskill/app/challenges/widget/presentation/ChallengeWidgetReducer.kt#L189-L213

We don't have `Flow` that notifies about `Project` and `Stage` completion, so it's done via step solved callback with performing API request